1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,611 @@
+
+The output of a pipeline chain is the concatenation of all its pipelines,
+in the sequence of their output:
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Output 'd'
+$x # 'a','b','c','d'
+```
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Error 'Bad' || Write-Output 'd'
+# Writes the error record 'Bad'
+$x # 'a','b','c','d'
+```
+
+Commands that fail will have any pipeline output
+emitted before evaluation of the chain operator.
+
+For example:
+
+```powershell
+$x = cmd1 && cmd2
+```
+
+If `cmd1` fails but emits output, `$x` will hold that value.
+
+#### Error semantics
+
+Errors will have the same semantics as the equivalent
+`cmd1; if ($?) { cmd2 }` syntax.
+
+Terminating errors will terminate the entire pipeline chain.
+But output already emitted by earlier pipelines in the chain
+will be output by the chain before terminating:
+
+```powershell
+try
+{
+    $x = cmd1 && $(throw "Bad")
+}
+catch
+{
+}
+
+$x # Has values from cmd1
+```
+
+Non-terminating errors will cause the immediate pipeline to continue,
+and the pipeline chain will evaluate as normal based on the value of `$?`.
+
+### Other notes
+
+#### New Abstract Syntax Tree (AST) types
+
+The new AST type, `PipelineChainAst`,
+will be created to represent a pipeline chain.
+This will inherit from `PipelineBaseAst`
+and may occur anywhere a `PipelineBaseAst` does in a PowerShell AST.
+
+`ICustomAstVisitor2` and `AstVisitor2` would be extended to deal with this AST,
+and .NET Core 3's new default interface implementation feature would be
+leveraged to ensure this does not break things
+as previous syntactic introductions have been forced to.
+
+#### Statements may not be chained
+
+PowerShell has a notable divergence from POSIX shells:
+
+- In POSIX shells, a pipeline is composed of statements
+- In PowerShell, a pipeline is a kind of statement
+
+This means that bash allows the following:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF'
+```
+
+In fact, this can be evaluated in the background:
+
+```bash
+if [ -n $VAR ]; then echo 'IF'; fi || echo 'AFTER-IF' &
+```
+
+However, in PowerShell, pipelines are subordinate to statements.
+The proposed implementation of chain operators precludes use of
+`&&` and `||` between statements like `if` or `while`,
+or constructions like `$x = cmd1 && $y = cmd2 && $x + $y`.
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+## Related Material
+
+- Original issue: [PowerShell/PowerShell #3241](https://github.com/PowerShell/PowerShell/issues/3241).
+
+- Work-in-progress implementation: [PowerShell/PowerShell #9849](https://github.com/PowerShell/PowerShell/pull/9849).
+
+- [Current handling of `&&` and `||` in the parser](https://github.com/PowerShell/PowerShell/blob/af1de9e88d28014438ff3414e82298e5b14f6e81/src/System.Management.Automation/engine/parser/Parser.cs#L5846-L5860).
+
+## Alternate Proposals and Considerations
+
+### `&&` and `||` as a statement separator

```

Comment:
We’re talking about making Bash users easier to adapt, but the limitations mentioned in this section worry me a lot. Most likely this will be misleading Bash users and we will receive a lot of negative feedback and requests to implement it.
Bash users will see these operators, but will find that they do "not work as expected". This is the worst thing that can happen. It is like implementing operator + which "does not always work as expected" (0+1=1 but 1+0=0).

Although I like this proposal in general, I would not want to accept it in this form and create a headache for Bash users.

I see two ways - either try to get as close as possible to bash or act like with the name: we have already chosen another name - pipeline chains, which means we can choose another implementation - maybe use &&& and ||| that should indicate that it works almost like && and || but a little different.

@rjmholt Great work!

@mklement0 Please review the RFC.

Reply:
@iSazonov Glad to see that I'm not the only one with those concerns.

Reply:
Please no `&&&` and `|||`